### PR TITLE
Fix incorrect parsing of `#` in CsvCommands, add missing tests for the spec-compliant behavior

### LINKED ISSUE #13907: Import-Csv / ConvertFrom-Csv misinterpret a double-quoted hash  (#) as starting a comment line
Follow-up from #13857:

In #2480, a deliberate enhancement was introduced for ignoring one or more `#`-prefixed comment (metadata) lines at the start of data passed to `Import-Csv` / `ConvertFrom-Csv`, so as to support common logging file formats.

However, the **detection of such comment lines is overzealous in that even a _double-quoted_ `#` (`"#"` as opposed to `#`) as the first non-whitespace token at the start of the line is misidentified as a comment**.

For instance, the following valid CSV data is mishandled:

```
"#Tag","Name"
"10","foo"
```

(An _unquoted_ `#` at the start of the line can similarly be part of a valid header row, though in that case there is unavoidable ambiguity, and defaulting to interpretation as a comment line, as is already the case, is defensible. Fortunately, `Export-Csv` and `ConvertTo-Csv` _always_ double-quote fields by default.)

## Steps to reproduce

```powershell
[pscustomobject]@{ '#Tag' = 10; Name = 'foo' } | ConvertTo-Csv | ConvertFrom-Csv | 
  ConvertTo-Json -compress | Should -Be '{"#Tag":"10","Name":"foo"}'
```

## Expected behavior

The test should succeed due to successful round-trip conversion to and from CSV.

## Actual behavior

The test fails, because the first line is misidentified as a comment, despite the `#` being double-quoted.

```none
Expected '{"#":"10","Name":"foo"}', but got $null.
```

## Environment data

```none
PowerShell Core 7.1.0-rc.2
```

### LINKED ISSUE #26419: Import-Csv and ConvertFrom-Csv don't parse correctly when a header contains `"#"`
### Prerequisites

- [x] Write a descriptive title.
- [x] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [x] Search the existing issues.
- [x] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [x] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

PowerShell does not parse CSVs correctly when one of the header rows contains a quoted hashtag value like `"#"`. 

I believe that this is incorrect behavior caused by the support added for [comments in the W3C Extended Log format](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/import-csv?view=powershell-7.6#:~:text=Import%2DCsv%20also,File%20Format.) in #2482:

> Lines starting with the hash character (#) are treated as comments and ignored

It incorrectly assumes that the `#` is a comment, but it's actually a header value in a quoted string. This is not following the spec as the line didn't start with a `#` character, but a `"` character.

I don't see any tests for W3C Extended Log format comments in CSVs in [Import-Csv.tests.ps1](test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Csv.Tests.ps1) or [ConvertFrom-Csv.tests.ps1](test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-Csv.Tests.ps1) - neither the happy path nor this edge case are present.

I believe that the incorrect parsing is happening here:
https://github.com/PowerShell/PowerShell/blob/8b7e493dd9e2e6d020509ec20c14cd3ba7208e14/src/Microsoft.PowerShell.Commands.Utility/commands/utility/CsvCommands.cs#L1337-L1368

`ParseNextRecord(values, builder)` converts the line `"#","Player","Score"` into `@("#", "Player", "Score")`, which strips the quotes in the header. Then, later, the parsed values are checked to see if the first value is a `#`. This is different from checking if the line starting with `#`, because we stripped the beginning character already.

I think that this check for a line starting with `#` should happen on the original string representing the line, not on the parsed values.

---

Here's a short script to set up the failing condition

```powershell
$fileContent = @'
"#","Player","Score"
"1","Kyle","2"
"2","Kirsten","5"
'@
Set-Content "./test.csv" -Value $fileContent
```

### Expected behavior

```console
Import-Csv "./test.csv"

# Player  Score
- ------  -----
1 Kyle    2
2 Kirsten 5

ConvertFrom-Csv $fileContent

# Player  Score
- ------  -----
1 Kyle    2
2 Kirsten 5
```

### Actual behavior

```console
Import-Csv "./test.csv"

1 Kyle    2
- ----    -
2 Kirsten 5

ConvertFrom-Csv $fileContent

1 Kyle    2
- ----    -
2 Kirsten 5
```

### Error details

```console

```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.5.4
PSEdition                      Core
GitCommitId                    7.5.4
OS                             Microsoft Windows 10.0.26200
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

_No response_

### PR DESCRIPTION
**Pull request by kylejrp**
# PR Summary

Fixes #26419
Fixes #13907

This pull request improves the handling of CSV files containing lines starting with `#` or `#Fields:` in both the `Import-Csv` and `ConvertFrom-Csv` cmdlets. The main focus is to correctly distinguish between comment lines, header lines, and quoted fields containing `#`, ensuring compatibility with the W3C Extended Log File Format. This is accomplished by modifying the `ReadHeader()` method of `CsvCommands` to check if the raw line starts with a `#`, instead of checking if the first parsed header value starts with a `#`. The changes also include new unit tests to verify these behaviors in `Import-Csv` and `ConvertFrom-Csv` cmdlets.

## PR Context

This pull requests fix a bug pointed out by #26419, which is that this CSV incorrectly gets its header row skipped, despite not starting with a `#` character.

```
"#","b","c"
"1","2","3"
"4","5","6"
```

Before, this would skip the header row of "#","b","c" and make "1","2","3" the header, with the first row having values of "1"="4", "2"="5","3"="6". Now, the first row will be correctly parsed as a header, and the first row will correctly be "#"="1", "b"="2", "c"="3", with the second row being "#"="4", "b"="5", "c"="6".

The original implementation broke the [public contract documented on the Import-Csv page](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.utility/import-csv?view=powershell-7.6#:~:text=Import%2DCsv%20also,File%20Format.), which follows the [W3C Extended Log File Format](https://www.w3.org/TR/WD-logfile.html):

> Lines starting with the hash character (`#`) are treated as comments and ignored unless the comment starts with `#Fields:` and contains delimited list of column names. In that case, the cmdlet uses those column names.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
  - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge. If this PR is a work in progress, please open this as a [Draft Pull Request and mark it as Ready to Review when it is ready to merge](https://docs.github.com/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests#draft-pull-requests).
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
  - [x] None
  - **OR**
  - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.5/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
    - [ ] Experimental feature name(s): 
- **User-facing changes**
  - [x] Not Applicable
  - **OR**
  - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - [ ] Issue filed: 
- **Testing - New and feature**
  - [ ] N/A or can only be tested interactively
  - **OR**
  - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)

### CODE REVIEW

**Comment by kylejrp** [comment]
@microsoft-github-policy-service agree

**Comment by iSazonov** [comment]
This parser has been optimized to work as fast as possible. The approach you are using will slow down the parser too much. This cannot be accepted.
(Perhaps the best approach would be to pre-check the first two characters in the file (for quote and #)).

**Comment by kylejrp** [comment]
@iSazonov took your approach and fixed it in 495c536135f0db92bf71e082ebeee3d3618cfc9e. Does it look okay now?

It only has to peek the first character to see if the line starts with a `#` character, and then has the same behavior as before (`values[0] = values[0].Substring(9); Header = values;` if it's a `#Fields: ` directive, line is ignored otherwise). This implementation should have the same performance as the parser code before my PR.

**Comment by kylejrp** [comment]
Hmm, there may be a bug with existing behavior before my PR - the test is failing with this data:

```
$testColumnsWithFields = @'
#Fields: a b c
1,2,3
'@
```

When I run it with the example CSV from the W3C spec on the latest PowerShell, it returns a single combined header:

```
PS C:\Users\kylej> $testData = @'
>> #Version: 1.0
>> #Date: 12-Jan-1996 00:00:00
>> #Fields: time cs-method cs-uri
>> 00:34:23 GET /foo/bar.html
>> 12:21:16 GET /foo/bar.html
>> 12:45:52 GET /foo/bar.html
>> 12:57:34 GET /foo/bar.html
>> '@
PS C:\Users\kylej> $testData | ConvertFrom-Csv

time cs-method cs-uri
---------------------
00:34:23 GET /foo/bar.html
12:21:16 GET /foo/bar.html
12:45:52 GET /foo/bar.html
12:57:34 GET /foo/bar.html

PS C:\Users\kylej> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.5.4
PSEdition                      Core
GitCommitId                    7.5.4
OS                             Microsoft Windows 10.0.26200
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

I expected `$testData` to look like this in the current PowerShell:

```
PS C:\Users\kylej> $testData | ConvertFrom-Csv

time     cs-method cs-uri
----     --------- ------
00:34:23 GET       /foo/bar.html
12:21:16 GET       /foo/bar.html
12:45:52 GET       /foo/bar.html
12:57:34 GET       /foo/bar.html
```

I can *almost* get it to work with this:

```
PS C:\Users\kylej> $testData | ConvertFrom-Csv -Delimiter ' '

00:34:23 GET /foo/bar.html
-------- --- -------------
12:21:16 GET /foo/bar.html
12:45:52 GET /foo/bar.html
12:57:34 GET /foo/bar.html
```

...but then it skips the header row.

Did support for W3C Extended Log File Format break at some point?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/CsvCommands.cs [hunk 1]
```diff
@@ -1336,35 +1336,32 @@ internal void ReadHeader()
...
-                else if (values.Count != 0 && values[0].StartsWith('#'))
-                {
-                    // Skip all lines starting with '#'
-                }
-                else
-                {
-                    // This is not W3C Extended Log File Format
-                    // By default first line is Header
-                    Header = values;
+                    if (values.Count != 0 && values[0].StartsWith("#Fields: "))
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/CsvCommands.cs]: We parse record and then do a check for `#Fields:` but we can do the check and only then parse if needed.

**Comment by kylejrp** [comment]
Given that the behavior in PowerShell today is to assume that W3C ELFF files are comma delimited (despite commas not being valid delimiters by the spec), I've updated the two failing test cases in bd2a8336829760cc5561295f6dc04a620e6f7000 to use commas as their delimiter. Tests should (hopefully) pass now.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Csv.Tests.ps1 [hunk 2]
```diff
@@ -182,3 +182,59 @@ Describe "Import-Csv with different newlines" -Tags "CI" {
...
+        $testColumnsWithUnquotedHashtag = @'
+#,b,c
+1,2,3
+4,5,6
+'@
+
+        $testColumnsWithFields = @'
+#Fields: a b c
+1,2,3
+'@
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Csv.Tests.ps1:218]: It is not work because delimiter in header is a space, in data is comma. They should be the same.
Current code doesn't detect the delimiter and use default one or from Delimiter parameter.
So we should use `#Fields: a,b,c` in the test.

Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Csv.Tests.ps1:218]: I see you already fixed this.

Comment (kylejrp) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Csv.Tests.ps1:218]: Fixed, but the tests are still failing. Seems like this behavior is broken in the current version of PowerShell:

```
PS C:\Users\kylej> $testData = @'
>> #Fields: a,b,c
>> 1,2,3
>> '@
PS C:\Users\kylej> ConvertFrom-Csv $testData
```

It fails to parse correctly. This is also true of space-delimited files. See discussion in #26428.

**Comment by kylejrp** [comment]
Added failing tests in 58948f0 to cover the space-delimited `#Fields: ` file case. Note that both the space-delimited and comma-delimited W3C ELFF files are currently broken in the latest PowerShell (see #26428), so both the `#Fields: ` tests will fail until we fix that here.

```
PS C:\Users\kylej> $testData = @'
>> #Fields: a,b,c
>> 1,2,3
>> '@
PS C:\Users\kylej> ConvertFrom-Csv $testData
PS C:\Users\kylej> $testData = @'
>> #Fields: a b c
>> 1 2 3
>> '@
PS C:\Users\kylej> ConvertFrom-Csv $testData -Delimiter ' '
PS C:\Users\kylej> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.5.4
PSEdition                      Core
GitCommitId                    7.5.4
OS                             Microsoft Windows 10.0.26200
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by kylejrp** [comment]
@iSazonov discovered some more failing edge cases, added them as `-Pending` tests to be solved by #26428. These non-`-Pending` tests should cover what works in PowerShell today.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-Csv.Tests.ps1 [hunk 3]
```diff
@@ -72,11 +72,101 @@ Describe "ConvertFrom-Csv DRT Unit Tests" -Tags "CI" {
...
+
+        $testColumnsWithUnquotedHashtag = @'
+#,b,c
+1,2,3
+4,5,6
+'@
+
+        $testColumnsWithFields = @'
+#Version: 1.0
+#Fields: a,b,c
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-Csv.Tests.ps1:99]: Please use full and correct W3C header in test. We have no need to support or detect broken header.

Comment (kylejrp) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-Csv.Tests.ps1:99]: But this is a correct and valid W3C header? It's not an invalid or broken header. `#Version: ` and `#Fields: ` can come in any order, and the other directives like `#Software: `, `#Start-Date: `, `#Date: `, and `#Remark: ` aren't required:

> An extended log file contains a sequence of lines containing ASCII characters terminated by either the sequence LF or CRLF. Log file generators should follow the line termination convention for the platform on which they are executed. Analyzers should accept either form. **Each line may contain either a directive** or an entry.

> Lines beginning with the # character contain directives. The following directives are defined:
> Version: \<integer\>.\<integer\>
> Fields: [\<specifier\>...]
> Software: string
> Start-Date: \<date\> \<time\>
> End-Date:\<date\> \<time\>
> Date:\<date\> \<time\>
> Remark: \<text\>

> **The directives Version and Fields are required and should precede all entries in the log.** The Fields directive specifies the data recorded in the fields of each entry.

What makes you think that this is an invalid W3C header? Are you able to help me understand by referencing the [W3C ELFF spec](https://www.w3.org/TR/WD-logfile.html)?

Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-Csv.Tests.ps1:99]: Sorry, I didn't see the commit where you disabled tests, and I thought you disabled this test.

Comment (kylejrp) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-Csv.Tests.ps1:99]: I don't understand where you got confused - as far as I can tell, the tests have only used valid headers so far across commits. But I'm glad that we're on the same page now.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-Csv.Tests.ps1 [hunk 4]
```diff
@@ -72,11 +72,101 @@ Describe "ConvertFrom-Csv DRT Unit Tests" -Tags "CI" {
     It "Test ConvertFrom-Csv with pipelined InputObject and Header" {
         $inputObject = [pscustomobject]@{ First = 1; Second = 2 }
         $res = $inputObject | ConvertTo-Csv
-        $result = $res | ConvertFrom-Csv -Header "Header1","Header2"
+        $result = $res | ConvertFrom-Csv -Header "Header1", "Header2"
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-Csv.Tests.ps1]: Please remove all unrelated formatting changes. (All above in tests.)

**Review by iSazonov — COMMENTED** [review]
LGTM with one style comment.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-Csv.Tests.ps1 (hunk 4 above)
Comment (kylejrp) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-Csv.Tests.ps1]: Fixed

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Csv.Tests.ps1 [hunk 5]
```diff
@@ -182,3 +182,111 @@ Describe "Import-Csv with different newlines" -Tags "CI" {
...
+
+Describe "Import-Csv containing #" -Tags "CI" {
+    BeforeAll {
+        $testColumnsWithQuotedHashtag = @'
+"#","b","c"
+1,2,3
+'@
+
+        $testColumnsWithUnquotedHashtag = @'
+#,b,c
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Import-Csv.Tests.ps1:194]: Ideally this should work like previous test. I suggest add such comment in the test below until we fix header parsing for full W3C support.

**Comment by muffy86** [comment]
This pull request appears inactive. If there are plans to continue work, please update the PR within 14 days; otherwise it may be closed.
